### PR TITLE
#5190 - Add metadata_modified field to resource

### LINKED ISSUE #5190: Last modified dates not updated for resources
### Please describe the expected behaviour

Updating metadata fields (title, description, etc) for resources should update the _metadata last updated_ attribute of the resource.

### Please describe the actual behaviour

_Metadata last updated_ value remains the same after edit

### What steps can be taken to reproduce the issue? 

* Update `package/resource_read.html` to show full localized datetime for the "Metadata last updated" field
```
              <tr>                
                <th scope="row">{{ _('Data last updated') }}</th>
                <td>{{ h.render_datetime(res.last_modified, "%c", False) or h.render_datetime(res.created, "%c", False) or _('unknown') }}</td>
              </tr>
              <tr>                            
                <th scope="row">{{ _('Metadata last updated') }}</th>
                <td>{{ h.render_datetime(res.revision_timestamp, "%c", False) or h.render_datetime(res.created, "%c", False) or _('unknown') }}</td>                     
              </tr>    
```
* Create resource
* Update metadata field
* Check to see if `last_modified` field is updated to current time

On create:
> Screenshot from 2020-02-08 09-35-59

On update:
> Screenshot from 2020-02-08 09-36-27

### Proposed solutions

I've explored two potential solutions:

#### 1. Controller-level

In package controller https://github.com/ckan/ckan/blob/master/ckan/controllers/package.py#L586
```python
            try:
                if resource_id:
                    data['id'] = resource_id
                    if resource_updated(context, data): # FIXME implement
                      data['last_modified'] = datetime.datetime.now()
                   get_action('resource_update')(context, data)
                else:
                    get_action('resource_create')(context, data)
```

#### 2. Dictization-level

Another approach might be when preparing to write the new data dictionary to the database, check if there are changes, and add the field here.

For example (in `resource_dict_save` https://github.com/ckan/ckan/blob/master/ckan/lib/dictization/model_save.py#L17):

```python
    for key, value in res_dict.iteritems():
        if isinstance(value, list):
            continue
        if key in ('extras', 'revision_timestamp', 'tracking_summary'):
            continue
        if key in fields:
            if isinstance(getattr(obj, key), datetime.datetime):
                if isinstance(value, string_types):
                    db_value = getattr(obj, key).isoformat()
                else:
                    db_value = getattr(obj, key)
                if  db_value == value:
                    continue
                if key == 'last_modified' and not new:
                    obj.url_changed = True
            if key == 'url' and not new and obj.url != value:
                obj.url_changed = True

            # keep track of fields we are updating
            if getattr(obj, key) != value:
                updated_fields_count += 1 
            setattr(obj, key, value)
        else:
            # resources save extras directly onto the object, instead
            # of in a separate extras field like packages and groups
            new_extras[key] = value
    
    if updated_fields_count > 0:
        obj.last_modified = datetime.datetime.now()
    obj.state = u'active'
    obj.extras = new_extras

    session.add(obj)
    return obj
```

### PR DESCRIPTION
**Pull request by pdelboca**
Fixes #5190 

### Proposed fixes:
This PR adds a metadata_modified field to the Resource model to keep track of changes. It is updated in `resource_dict_save` (`model_save.py`) only if new values being saved differs from the actual ones.

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

**Comment by pdelboca** [comment]
@amercader regarding new unit tests, I'm thinking on what's the best way do do testing of this. In our codebase we have several ways to do it for the field in package:

`test_model_dictize.py::test_package_dictize_basic`:
```python
today = datetime.date.today().strftime("%Y-%m-%d")
assert result["metadata_modified"].startswith(today)
```

`test_get.py::TestPackageShow::test_package_show_with_full_dataset`
```python
def replace_datetime(dict_, key):
            assert key in dict_
            dict_[key] = u"2019-05-24T15:52:30.123456"

replace_datetime(dataset2, "metadata_modified")
```

`test_jobs.py::TestDictizeJob::test_dictize_job`
```python
assert abs((now - dt).total_seconds()) < 10
```

@smotornyuk proposed in a PR (can't find it) to use [freezegun](https://github.com/spulec/freezegun)  but it mocks `datetime.now()` and we are using `datetime.utcnow()`.

Creating a new mock is [not that trivial](https://stackoverflow.com/questions/4481954/trying-to-mock-datetime-date-today-but-not-working), since Python built-in types are immutable. I tried to create a fixture like the following one proposed [here](https://stackoverflow.com/questions/20503373/how-to-monkeypatch-pythons-datetime-datetime-now-with-py-test):
```python
import datetime
import pytest

FAKE_TIME = datetime.datetime(2020, 12, 25, 17, 5, 55)

@pytest.fixture
def patch_datetime_utcnow(monkeypatch):

    class mydatetime:
        @classmethod
        def utcnow(cls):
            return FAKE_TIME

    monkeypatch.setattr(datetime, 'datetime', mydatetime)


def test_patch_datetime(patch_datetime_now):
    assert datetime.datetime.now() == FAKE_TIME
```

But it throws errors on other tests and codebase because we are mocking the `datetime.datetime` object that it is widely used. For example:

```
ERROR    ckan.lib.search.common:common.py:63 mydatetime() takes no arguments
Traceback (most recent call last):
  File "/home/pdelboca/Repos/ckan/ckan/lib/search/common.py", line 61, in is_available
    conn.search(q="*:*", rows=1)
  File "/home/pdelboca/envs/ckan-py3/lib/python3.7/site-packages/pysolr.py", line 721, in search
    decoded = self.decoder.decode(response)
  File "/home/pdelboca/envs/ckan-py3/lib/python3.7/site-packages/simplejson/decoder.py", line 374, in decode
    obj, end = self.raw_decode(s)
  File "/home/pdelboca/envs/ckan-py3/lib/python3.7/site-packages/simplejson/decoder.py", line 404, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
  File "/home/pdelboca/Repos/ckan/ckan/lib/search/common.py", line 101, in solr_datetime_decoder
    date_values['second'])
TypeError: mydatetime() takes no arguments
```

I can keep testing as we are doing now, but I thought that it was good to raise this and found a holistic approach for how we test this time based logic.

**Comment by pdelboca** [comment]
@amercader regarding new unit tests, I'm thinking on what's the best way to do testing of this. In our codebase we have several ways to do it for the field in package:

`test_model_dictize.py::test_package_dictize_basic`:
```python
today = datetime.date.today().strftime("%Y-%m-%d")
assert result["metadata_modified"].startswith(today)
```

`test_get.py::TestPackageShow::test_package_show_with_full_dataset`
```python
def replace_datetime(dict_, key):
            assert key in dict_
            dict_[key] = u"2019-05-24T15:52:30.123456"

replace_datetime(dataset2, "metadata_modified")
```

`test_jobs.py::TestDictizeJob::test_dictize_job`
```python
assert abs((now - dt).total_seconds()) < 10
```

@smotornyuk proposed in a PR (can't find it) to use [freezegun](https://github.com/spulec/freezegun)  but it mocks `datetime.now()` and we are using `datetime.utcnow()`.

Creating a new mock is [not that trivial](https://stackoverflow.com/questions/4481954/trying-to-mock-datetime-date-today-but-not-working), since Python built-in types are immutable. I tried to create a fixture like the following one proposed [here](https://stackoverflow.com/questions/20503373/how-to-monkeypatch-pythons-datetime-datetime-now-with-py-test):
```python
import datetime
import pytest

FAKE_TIME = datetime.datetime(2020, 12, 25, 17, 5, 55)

@pytest.fixture
def patch_datetime_utcnow(monkeypatch):

    class mydatetime:
        @classmethod
        def utcnow(cls):
            return FAKE_TIME

    monkeypatch.setattr(datetime, 'datetime', mydatetime)


def test_patch_datetime(patch_datetime_now):
    assert datetime.datetime.now() == FAKE_TIME
```

But it throws errors on other tests and codebase because we are mocking the `datetime.datetime` object that it is widely used. For example:

```
ERROR    ckan.lib.search.common:common.py:63 mydatetime() takes no arguments
Traceback (most recent call last):
  File "/home/pdelboca/Repos/ckan/ckan/lib/search/common.py", line 61, in is_available
    conn.search(q="*:*", rows=1)
  File "/home/pdelboca/envs/ckan-py3/lib/python3.7/site-packages/pysolr.py", line 721, in search
    decoded = self.decoder.decode(response)
  File "/home/pdelboca/envs/ckan-py3/lib/python3.7/site-packages/simplejson/decoder.py", line 374, in decode
    obj, end = self.raw_decode(s)
  File "/home/pdelboca/envs/ckan-py3/lib/python3.7/site-packages/simplejson/decoder.py", line 404, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
  File "/home/pdelboca/Repos/ckan/ckan/lib/search/common.py", line 101, in solr_datetime_decoder
    date_values['second'])
TypeError: mydatetime() takes no arguments
```

I can keep testing as we are doing now, but I thought that it was good to raise this and found a holistic approach for how we test this time based logic.

**Comment by pdelboca** [comment]
@amercader discard the last comment, writing it just make me read more carefully the docs and issues and seems that `freezegun` already mocks `utcnow()` as the `README.md` points out:

> Once the decorator or context manager have been invoked, all calls to datetime.datetime.now(), datetime.datetime.utcnow(), datetime.date.today(), time.time(), time.localtime(), time.gmtime(), and time.strftime() will return the time that has been frozen.

Sorry, holidays are coming soon :) 

I will add some tests using `freezegun` and see how it goes!

**Comment by wardi** [comment]
If we add these fields then we can stop updating the `metadata_modified` at the package level in the database and instead replace the `metadata_modified` value returned from package_show with `max(package.metadata_modified, resource1.metadata_modified, ...)` That might help reduce the contention raised in #5233

File: ckan/model/resource.py [hunk 1]
```diff
@@ -47,6 +48,7 @@
     Column('size', types.BigInteger),
     Column('created', types.DateTime, default=datetime.datetime.utcnow),
     Column('last_modified', types.DateTime),
+    Column('metadata_modified', types.DateTime),
```
[the remark this answers is not in the document]
Comment (amercader) [reply ckan/model/resource.py]: Ah, I see now that the field will be set to utcnow as the value of the field will be different, but I'd add it here anyway for consistency

File: ckan/lib/dictization/model_save.py [hunk 2]
```diff
@@ -56,12 +57,16 @@ def resource_dict_save(res_dict, context):
                     obj.url_changed = True
             if key == 'url' and not new and obj.url != value:
                 obj.url_changed = True
+            if getattr(obj, key) != value:
```
Comment (amercader) [inline ckan/lib/dictization/model_save.py]: If the `key` was not present in the resource before this will raise an `AttributeError`.

Comment (amercader) [reply ckan/lib/dictization/model_save.py]: Sorry ignore me, I was reading the code wrong

**Comment by amercader** [comment]
@pdelboca I went ahead and added the missing bits to this before merging. More important ones:

* 873199fb6 [#5190] Handle custom fields in metadata_modified logic
* 0e0a67b13 [#5190] Set metadata_modified to created on existing resources

**Comment by amercader** [comment]
@wardi this PR only included the addition of the `metadata_modified` field at the resource level. The dataset one remains unchanged. I guess that to do what you propose we need similar logic on model_save at the package level to see if the dataset really changed

**Pull request merged**